sw/device/lib/dif/dif_spi_device.c

```diff
@@ -0,0 +1,695 @@
+  // we've read so far. For sending, we advance the write pointer, which
+  // indicates how far ahead we've written.
+  fifo_ptr_t *ptr;
+  if (is_recv) {
+    ptr = &fifo->read_ptr;
+  } else {
+    ptr = &fifo->write_ptr;
+  }
+
+  uint16_t total_bytes_available = bytes_left;
+
+  // First, bring the fifo write pointer into word alignment, so we can do
+  // simple full-word I/O rather than partial word I/O; note that the FIFO
+  // SRAM is not guaranteed to support subword writes.
+  ptrdiff_t misalignment = misalignment32_of(ptr->offset);
+  if (misalignment != 0) {
+    // The number of bytes missing to bring `ptr->offset` back into alignment.
+    // For exmaple, 0x3 has misalignment of 3 and realignment of 1.
+    ptrdiff_t realignment = sizeof(uint32_t) - misalignment;
+    // Note that we might be doing less I/O than the misalignment requires; we
+    // might be off by a single byte, but not have the full three bytes for full
+    // realignment.
+    if (realignment > bytes_left) {
+      realignment = bytes_left;
+    }
+
+    // Converts `offset`, which points to a subword boundary, to point to the
+    // start of the current word it points into.
+    ptrdiff_t current_word_offset = ptr->offset - misalignment;
+    uint32_t current_word =
+        mmio_region_read32(spi->base_addr, fifo_base + current_word_offset);
+
+    // Act on only to a suffix of `current_word`, corresponding to the necessary
+    // realignment.
+    uint8_t *current_subword = ((uint8_t *)&current_word) + misalignment;
+    if (is_recv) {
+      memcpy(byte_buf, current_subword, realignment);
+    } else {
+      // When sending, we need to commit the staged write.
+      memcpy(current_subword, byte_buf, realignment);
+      mmio_region_write32(spi->base_addr, fifo_base + current_word_offset,
+                          current_word);
+    }
+
+    fifo_ptr_increment(ptr, realignment, fifo_len);
+    byte_buf += realignment;
+    bytes_left -= realignment;
+  }
+
+  // Now, we just do full word I/O until we run out of stuff to act on.
+  while (bytes_left > 0) {
+    // At the end, we may not have a full word to copy, but it's otherwise
+    // the same case as a full word, since we're already word aligned (if
+    // this would be a subword read, it would end the loop anyway).
+    uint16_t bytes_to_copy = bytes_left;
+    if (bytes_to_copy > sizeof(uint32_t)) {
+      bytes_to_copy = sizeof(uint32_t);
+    }
+
+    uint32_t current_word = 0;
+    if (is_recv || bytes_to_copy != sizeof(uint32_t)) {
+      // If recieving, we need to read this word always.
+      //
+      // If sending, we only need to write a prefix when writing a subword.
+      // In that case, we need to avoid clobbering the current contents of the
+      // FIFO.
+      current_word =
+          mmio_region_read32(spi->base_addr, fifo_base + ptr->offset);
+    }
+
+    // Copy a prefix; most of the time, this will be the whole word.
```

Comment:
Unless I'm missing something, we can refactor lines 598-610 and 634-646 to a function to eliminate this duplication.

Reply:
I mean, I guess? You'd need to pass like, seven arguments though, and I worry that would obscure meaning too much for what is a very tiny snippet.

Reply:
Good point :)